### DISCUSSION TITLE
Key migration from MINIO_KMS_MASTER_KEY to kes

### DISCUSSION THREAD

**Discussion by nowheresly** [minio/minio#15644] [Q&A] [answered]
Hi,

I am trying to migrate a minio cluster using sse-s3 from the now deprecated environment variable MINIO_KMS_MASTER_KEY to a regular key stored in kes.

So previously, we add an environment variable configured as follow:

`MINIO_KMS_MASTER_KEY="encryptkey:cafebabecafebabecafebabecafebabecafebabecafebabecafebabecafebabe"
`

Please not that the key is encoded in an hexadecimal form and the name of the key is **encryptkey**.

Now, the minio cluster uses KES with this configuration:

```
      MINIO_KMS_KES_ENDPOINT=https://kesserver:7373
      MINIO_KMS_KES_KEY_FILE=client.key
      MINIO_KMS_KES_CERT_FILE=client.crt
      MINIO_KMS_KES_CA_PATH=public.crt
      MINIO_KMS_KES_KEY_NAME: encryptkey
```
The KES server is deployed using the filesystem keystore as described here: https://github.com/minio/kes/wiki/Filesystem-Keystore

On startup, the KES server is creating an "encryptkey" file in its keystore.

Now here is my question: I can see the "encryptkey" file of KES that the key is no more stored in hexadecimal, but rather in base64.

`{"bytes":"yv66vsr+ur7K/rq+yv66vsr+ur7K/rq+yv66vsr+ur4=","algorithm":"AES256-GCM_SHA256","created_at":"2022-09-01T11:26:24.632595656Z","created_by":"6042a399bed6ccb5059f70628485e5aad8eb5ce5a4b6a6173df3f67d3956138f"}`

Since my key was encoded in hexa before, I converted it to base64 (using a tool like https://cryptii.com/pipes/hex-to-base64 for example)

* hexa: cafebabecafebabecafebabecafebabecafebabecafebabecafebabecafebabe
* base64: yv66vsr+ur7K/rq+yv66vsr+ur7K/rq+yv66vsr+ur4=

Unfortunately, this does not work and I get consistently this error when minio tries to decrypt an encrypted file using the key provided by KES:

`We encountered an internal error, please try again.: cause(decryption failed: ciphertext is not authentic)`

This error means the key is incorrect.

How can i convert my previous hexadecimal key in something that can be served by KES ?

The encryption algorithm has changed between both approaches ?

Am I doing something wrong ?

Thanks a lot for your help

**Comment by vadmeste** [comment]
I think that should work, /cc @aead 

@Nowheresly, did you upgrade MinIO from an older version ?

**Comment by aead** [comment, marked as answer]
@Nowheresly There is no migration path from `MINIO_KMS_MASTER_KEY` to KES. 
`MINIO_KMS_MASTER_KEY` is a legacy env. var. that was supposed to be used only for **testing** purposes. The encryption implementation for KES and `MINIO_KMS_MASTER_KEY` are not compatible.
 
(We noticed that several users ran MinIO using this env. var. Therefore, we introduced a **supported**  env. var. that implements a single key KMS **and** is compatible to KES: `MINIO_KMS_SECRET_KEY`.

So:
 - There is no upgrade path from `MINIO_KMS_MASTER_KEY` to `MINIO_KMS_SECRET_KEY` or KES.
 - There is a clean upgrade path from `MINIO_KMS_SECRET_KEY` to KES.

In cases where data is encrypted using `MINIO_KMS_MASTER_KEY`, unfortunately, the only option is to migrate the data from one cluster to another one - e.g. with `mc mirror`.

**Comment by nowheresly** [comment]
Hi @vadmeste , @aead 
thank you very much for taking the time to answer my question.
The explanation is very clear.
